### ISSUE TITLE
Thoughts on adding an equivalent of Flow's $PropertyType to give access to nested properties of a decoder?  

### ISSUE THREAD

**Issue by galencorey** [nvie/decoders#367]
In Flow, if you have a type `User`: 

```js
   type User = {|
      name: string, 
      age: number, 
      address: {
         street: string, 
         city: string, 
         state: string, 
         zip: number, 
      }
   |}
``` 
You can grab the address type by using $PropertyType utility type: 
```js
   type Address = $PropertyType<User, 'address'>
```

I have found this to be extremely useful in the case of using flow types that are generated from graphql queries, because it allows for better consistency in types throughout the code. My proposal is to add some similar functionality for decoders. If you had a similar decoder for that same user, looking like this: 

```js 
   import { object, number, string } from 'decoders';  

   const userDecoder = object({
      name: string, 
      age: number, 
      address: object({
         street: string, 
         city: string, 
         state: string, 
         zip: number, 
      })
   }); 
```
That you would be able to access the address in this way: 
```js
   import { propertyType } from 'decoders';  
   ...
   const addressDecoder = propertyType(userDecoder, 'address'); 
```

It would also be ideal to be able to access these even when nested within an array, similar to $ElementType. 

I would find this extremely useful with generated decoders, for the same reason that it is useful with flow types. I would love to get your thoughts on this (or maybe it is already possible!)

**Comment by lydell** [comment]
Hi! Interesting idea! I just randomly found this issue when checking things out, and became curious. Why not do this?

```js
import { object, number, string } from "decoders";

const addressDecoder = object({
  street: string,
  city: string,
  state: string,
  zip: number,
});

const userDecoder = object({
  name: string,
  age: number,
  address: addressDecoder,
});
```

**Comment by galencorey** [comment]
@lydell Thank you for your response! I agree that in most cases this is usually a great solution to the problem (it is also a good solution in Flow). Writing the subtypes individually and then composing them together is how I would handle it if I were writing the types/decoders myself. Maybe in a perfect world, all types/decoders would all be written like this.

The use case I have in mind is automatically generated types (in my case generated from GraphQl queries). I find it useful to be able to work backward from a parent type and pull out one of its "property types." I would find a similar feature in decoders to be useful for the same reason.

**Comment by lydell** [comment]
Ah. You mean that if you have some tool that automatically generates decoders you would want to extract part of it? (Is there such a tool, btw? Sounds interesting!)

Btw – when I worked with GraphQL a couple of years ago, we generated Flow types from all our queries using apollo-codegen. Then we annotated the GraphQL responses with those types (they were `any` by default). We just trusted the responses to really have the shapes of the generated types, and didn't bother decoding them. I never had an issue where the types didn’t match reality. GraphQL has some pretty strong guarantees, after all!

**Comment by nvie** [comment]
**tl;dr** I think the outer decoders API *could* theoretically support this technique, but in the current implementation it's going to be hard.

Interesting idea!  This is something that I have sometimes pondered about while writing the decoder internals.  What you want is almost the opposite of composing.

One problem currently is that a `Decoder` in the current implementation is a JavaScript function — a specific function that has the following characteristics:

- It takes _anything_ as input
- It returns either an `Ok` or `Err` value

Simplified a bit, the type is roughly this:

```javascript
type Decoder<T> = (blob: mixed) => Ok<T> | Err<string>;
```

**So all decoders are just functions that return values.**

An `object()` expression therefore returns a JS _function_:

```javascript
const person: Decoder<Person> = object({
  name: string,
  age: number,
})   // Returns a Decoder<...>: a function!

// Demonstrate it's just a function
person(123)  // Err value
person({ name: 'Santa', age: 199 })  // Ok value
```

Because you cannot "look into" a function implementation, this means that we lose the ability to introspect the "name/age structure" we used to build this function with.  So we cannot "grab" the name or age field decoders back out.

On this level, what's happening is very much like the following:

```javascript
function adder(x) {
  return y => x + y;
}

const addThree = adder(3);
addThree(5)   // 8
```

Once you built the `addThree` function, there's no way of telling that it was built using the `3` argument down the road.  You can use it, but not introspect it anymore.

So I don't think building a propertyType function is possible, currently.


### I don't think it's impossible in theory

Now, this limitation is mostly the result of an implementation detail:  I chose to represent a `Decoder` as a JS function.  This has many advantages, like being the most light-weight structure, no extra support structures needed for good TypeScript/Flow inference, and lower memory usage.

Perhaps instead of letting the decoders return a JS function, but instead a "description" of their structure somehow, then we could introspect that structure after the fact to obtain an inner decoder and reuse it separately.  But then you would need another layer that turns the decoder description into the actual JS function that is the `Decoder` as it exists currently.

I don't think this added complexity and runtime overhead is worth it, though.  But it's a fun thought experiment on expressivity.

**Comment by nvie** [comment]
### A similar use case

There would be another use case were introspection would be useful, and that is when rendering error messages.  Because of the way the `either` family of decoders get nested — for example `either3(a, b, c)` is really just a convenience helper for `either(a, either(b, c))`.  Each such nested decoder will also produce its own error messages that is unaware of its context, and the inner error messages "bubble up" and get assembled upwards by their parent decoders.

So if the inner `either(b, c)` call returns the following error message:

```
null
^^^^
Either:
- Must be B
- Must be C
```

Then this message "bubbles up" and the outer `either()` has to treat it as an opaque error message and display it as the "second leg":

```
null
^^^^
Either:
- Must be A
- Either:
  - Must be B
  - Must be C
```

That's why you see that nesting.  A long standing desire has been to render this case as:

```
null
^^^^
Either:
- Must be A
- Must be B
- Must be C
```

But this would require some "introspection", i.e. take a sneak peek into the sub decoder, and if it is an `either`, then combine the alternatives in the error message.  I haven't found a good non-hacky solution to this yet.
